lectures/zksnark.tex

```diff
@@ -0,0 +1,271 @@
+                \draw[arrow, blue] (add) -- (c);
+            \end{tikzpicture}
+            &
+            \begin{tikzpicture}
+                % Nodes
+                \node[var] (a) at (0, -1.5) {$a$};
+                \node[var] (b) at (2, -1.5) {$b$};
+                \node[gate] (mul) at (1, 0) {$\times$};
+                \node[var] (c) at (1, 1.5) {$c$};
+
+                % Arrows
+                \draw[arrow] (a) -- (mul);
+                \draw[arrow] (b) -- (mul);
+                \draw[arrow, blue] (mul) -- (c);
+            \end{tikzpicture}
+        \end{tabular}
+        \caption{Addition and Multiplication Gates}
+    \end{minipage}
+
+    \vspace*{1em}
+\end{figure}
+
+\begin{wraptable}{r}{0.4\textwidth}
+    \centering
+    \vspace{-1em}
+    \begin{tabular}{|c|c|c|}
+        \hline
+        \textbf{A} & \textbf{B} & \textbf{A AND B} \\
+        \hline
+        0 & 0 & 0 \\
+        \hline
+        0 & 1 & 0 \\
+        \hline
+        1 & 0 & 0 \\
+        \hline
+        1 & 1 & 1 \\
+        \hline
+    \end{tabular}
+    \caption{AND Gate Truth Table}
+    \label{tab:and-truth-table}
+    \vspace{1em}
+\end{wraptable}
+
+% --- Finish Writing diagrams ---
+
+The \textit{AND Gate Truth Table~\ref{tab:and-truth-table}} shows us the results we receive if 
+particular values are supplied to the gate. The main point here is that with this table, we can 
+verify the validity of logical statements. Boolean circuits receive an input vector of $\{0, 1\}$ 
+and resolve to true (1) or false (0); basically, they determine if the input values satisfy the 
+statement.
+
+We can do the same with \textbf{arithmetic circuits} to verify computations without excessive 
+verbosity because of binary arithmetic.
+
+Let's take a look at some examples of programs and how can we translate them to the arithmetic
+circuits.
+
+Very simple program with a multiplication.
+
+\begin{lstlisting}[numbers=none]
+    return a * b
+\end{lstlisting}
+
+This can be represented as a circuit with only one gate:
+
+\begin{equation}
+    r = a \times b
+\end{equation}
+
+The witness vector (solution vector) is $w = \{r, a, b\}$, for example: $\{6, 2, 3\}$. We 
+assume that the $a$ and $b$ are input values. 
+
+We can think of the "=" in the gate as an assertion, meaning that if $a \times b$ does not equal
+$r$, the assertion fails, and the input values don't resolve the circuit.
+
+How can we translate an if statement?
+
+\begin{lstlisting}[numbers=none]
+    if (a) {
+        return b * c
+    } else {
+        return b + c
+    }
+\end{lstlisting}
+
+We can express this logic in mathematical terms as follows: "If $a$ is true, compute 
+$b \times c$; otherwise, compute $b + c$." Only numerical expressions are allowed. Knowing that
+$\text{true} := 1$ and $\text{false} := 0$, we can transform it as follows:
+
+\begin{equation}
+    r = a \times (b \times c) + (1 - a) \times (b + c)
+\end{equation}
+
+$w = \{r, a , b, c\}$: $\{6, 1, 2, 3\}$, $\{5, 0, 2, 3\}$. \\
```

Comment:
Specify, "It seems that the witness $w$ should be of a form..."

Besides, I believe it is better to use a tuple notation to represent a witness (that is $(r, a,b,c)$ instead of $\\{r,a,b,c\\}$). After all, the witness is not a set but an ordered set of scalars (a vector, in a sense).